pkg/api/service/util_test.go

```diff
@@ -214,3 +215,58 @@ func TestNeedsHealthCheck(t *testing.T) {
 		},
 	})
 }
+
+func TestWarnings(t *testing.T) {
+	testcases := []struct {
+		name     string
+		template *api.Service
+		expected []string
+	}{
+		{
+			name:     "null",
+			template: nil,
+			expected: nil,
+		},
+		{
+			name: "no dup ports",
+			template: &api.Service{Spec: api.ServiceSpec{
+				Ports: []api.ServicePort{
+					{Protocol: api.ProtocolTCP, Port: 443},
+					{Protocol: api.ProtocolTCP, Port: 1443},
+				},
+				Type: api.ServiceTypeNodePort,
+			},
+			},
+			expected: nil,
+		},
+		{
+			name: "dup ports",
+			template: &api.Service{Spec: api.ServiceSpec{
+				Ports: []api.ServicePort{
+					{Name: "tcp-443", Protocol: api.ProtocolTCP, Port: 443},
+					{Protocol: api.ProtocolTCP, Port: 1443},
+					{Name: "udp-443", Protocol: api.ProtocolUDP, Port: 443},
+				},
+				Type: api.ServiceTypeNodePort,
+			},
+			},
+			expected: []string{
+				`spec.ports[2].port: service has duplicate ports: 443`,
```

Comment:
same port different protocol is legit

Reply:
I think we have validation on this

Reply:
https://github.com/kubernetes/kubernetes/issues/47249#issuecomment-587960741
I think this will cover the patch problem. 

Maybe the warningOncreation can be nil.

Reply:
it seems we should add it only on patches https://github.com/kubernetes/kubernetes/issues/105610#issuecomment-940454445 per @liggitt comments

Reply:
we only want to warn when there's a real issue and there's something actionable the user can do

for this, my understanding if there's only a real issue when:
1. the port is actually duplicated (same port AND protocol)
2. the port number is duplicated and client-side apply / strategic merge patch is being used to create or update the object

the warning should make it clear what they should do in response... for problem 2, the issue isn't the ports, it's client-side apply / strategic merge patch, right?

Reply:
1. is a validation error, we shouldn't do more,it already fails, no?
2. yes

Reply:
If we can warn about ambiguity BEFORE client-side-apply ruins it all, then yes.  But I think same-port, same-protocol, different targetPort is legit?

Reply:
Yeah, it's legitimate, the problem is not the ports, it's that client-side apply / strategic merge patch is being used on this data.

I think the warning is *if* you are using client-side apply to create (can we detect via the annotation?) or strategic merge patch to update (not sure we have visibility to that here) or maybe are putting duplicate ports in a list in an object that is managed by client-side apply (maybe can tell if the annotation is present), you should stop using client-side apply

Reply:
I may add the warning to the bad place.

```
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  containers:
  - name: nginx
    image: nginx:1.14.2
    ports:
    - containerPort: 80
      name: a
      protocol: TCP
      hostPort: 10080
    - containerPort: 80
      name: a1
      protocol: UDP
      hostPort: 10080
    - containerPort: 443
      name: b
      protocol: UDP
      hostPort: 20080
  - name: nginx-1
    image: nginx:1.14.2
    ports:
    - containerPort: 443
      protocol: UDP
      name: b1
      hostPort: 40080
    - containerPort: 443
      protocol: UDP
      name: b2
      hostPort: 30080
```
create a pod based on the yaml above.
```
    - containerPort: 443
      protocol: TCP
      name: b3
      hostPort: 30080
```
Add it at the end of the yaml and apply again.
```
[root@paco ~]# kubectl get pod/nginx -o yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Pod","metadata":{"annotations":{},"name":"nginx","namespace":"default"},"spec":{"containers":[{"image":"nginx:1.14.2","name":"nginx","ports":[{"containerPort":80,"hostPort":10080,"name":"a","protocol":"TCP"},{"containerPort":80,"hostPort":10080,"name":"a1","protocol":"UDP"},{"containerPort":443,"hostPort":20080,"name":"b","protocol":"UDP"}]},{"image":"nginx:1.14.2","name":"nginx-1","ports":[{"containerPort":443,"hostPort":40080,"name":"b1","protocol":"UDP"},{"containerPort":443,"hostPort":30080,"name":"b2","protocol":"UDP"},{"containerPort":443,"hostPort":30080,"name":"b3","protocol":"TCP"}]}]}}
  name: nginx
  namespace: default
spec:
  containers:
  - image: nginx:1.14.2
    imagePullPolicy: IfNotPresent
    name: nginx
    ports:
    - containerPort: 80
      hostPort: 10080
      name: a
      protocol: TCP
    - containerPort: 80
      hostPort: 10080
      name: a1
      protocol: UDP
    - containerPort: 443
      hostPort: 20080
      name: b
      protocol: UDP
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-hhlmn
      readOnly: true
  - image: nginx:1.14.2
    imagePullPolicy: IfNotPresent
    name: nginx-1
    ports:
    - containerPort: 443
      hostPort: 40080
      name: b1
      protocol: UDP
    - containerPort: 443
      hostPort: 30080
      name: b2
      protocol: UDP
  dnsPolicy: ClusterFirst
  restartPolicy: Always
  schedulerName: default-scheduler
```

You can see b3 port in annotation, but it is de-duped during patching.

Reply:
This is an example of a scenario where we need to add a warning.

Reply:
```
[root@paco ~]# cat patch.yaml
spec:
  template:
    spec:
      containers:
      - name: nginx-1
        ports:
        - containerPort: 443
          protocol: UDP
          name: b1
          hostPort: 40080
        - containerPort: 443
          protocol: UDP
          name: b2
          hostPort: 30080
        - containerPort: 443
          protocol: TCP
          name: b3
          hostPort: 30080

[root@paco ~]# kubectl --v=8  patch deployments.apps --type=strategic nginx --patch-file patch.yaml
I1024 13:01:40.864347    9338 request.go:1073] Request Body: {"spec":{"template":{"spec":{"containers":
[{"name":"nginx-1","ports":[{"containerPort":443,"hostPort":40080,"name":"b1","protocol":"UDP"},
{"containerPort":443,"hostPort":30080,"name":"b2","protocol":"UDP"},
{"containerPort":443,"hostPort":30080,"name":"b3","protocol":"TCP"}]}]}}}}

### deploy yaml after applying
        - image: nginx:1.14.2
          imagePullPolicy: IfNotPresent
          name: nginx-1
          ports:
          - containerPort: 443
            hostPort: 30080
            name: b3
            protocol: TCP
          - containerPort: 443
            hostPort: 30080
            name: b2
            protocol: UDP
          resources: {}

```

Reply:
> I think the warning is _if_ you are using client-side apply to create (can we detect via the annotation?) or strategic merge patch to update (not sure we have visibility to that here) or maybe are putting duplicate ports in a list in an object that is managed by client-side apply (maybe can tell if the annotation is present), you should stop using client-side apply

@liggitt warning according to the applied annotation is doable.
Or can we add the warning to `kubectl client`? The logic would be too special.

Or we can do something similar to #92437, just add a warning when `deleteMatchingEntries` remove an entry with merge key('port' or 'containerPort').

Reply:
> I think the warning is _if_ you are using client-side apply to create (can we detect via the annotation?) or strategic merge patch to update (not sure we have visibility to that here) or maybe are putting duplicate ports in a list in an object that is managed by client-side apply (maybe can tell if the annotation is present), you should stop using client-side apply

@liggitt warning according to the applied annotation is doable.
Or can we add the warning to `kubectl client`? The logic would be too special.

Or we can do something similar to #92437, just add a warning when `deleteMatchingEntries` remove an entry with merge key('port' or 'containerPort').

Reply:
Updated.

Warning only if it is an update and there are duplicated ports.
- service update
- podTemplate update for ds/deploy/job
- pod update is intended to be empty

https://github.com/kubernetes/kubernetes/blob/9184a7e43bf5f82847ef22d137eb0ae8b65bb29b/pkg/registry/core/pod/strategy.go#L148-L155